File: validation/equivalence_ratio_case.jl
using Arrhenius, NPZ, Test, LinearAlgebra
using SHA: sha256
include(joinpath(@__DIR__,"..","example","thermodynamics","equivalence_ratio.jl"))

"""Compare every native equivalence_ratio_calculation output against the
pinned Cantera 4 thermo/equivalenceRatio.py reference (NPZ + JSON produced by
validation/equivalence_ratio_case.py from the same gri30.yaml)."""
function validate_equivalence_ratio_case(mechanisms, references)
    mechanism = joinpath(mechanisms,"gri30.yaml")
    data = npzread(joinpath(references,"equivalence_ratio-gri30.npz"))
    text(key) = String(vec(UInt8.(data[key])))
    println("Checking against Cantera ",text("cantera_version_utf8"),
            " (commit ",text("source_commit_utf8"),")")
    gas = CreateSolution(mechanism)
    result = equivalence_ratio_calculation(gas)
    mixtures = split(text("mixture_names_utf8"),'\n')
    Xref, Yref, Tref, Pref = data["X"], data["Y"], data["T_K"], data["P_Pa"]
    # Mixtures recorded before the HP equilibration stay at the 300 K default;
    # later mixtures retain the burnt temperature on both sides.
    fresh = ("X_stoich_mole","X_stoich_mass","X_Z055","X_fresh_burnt_case")
    max_composition_error = 0.0
    @testset "Cantera 4 thermo/equivalenceRatio.py" begin
        @test text("cantera_version_utf8")[1] == '4'
        @test bytes2hex(sha256(read(mechanism))) == text("mechanism_sha256_utf8")
        @test gas.species_names == split(text("species_names_utf8"),'\n')
        @test gas.n_species == 53
        for (i,name) in enumerate(mixtures)
            @testset "$name" begin
                Xnative = getproperty(result,Symbol(name))
                if name == "X_burnt"
                    # Native equilibrium tolerances; see validation/equilibrium_cases.jl.
                    @test result.T_burnt_K ≈ Tref[i] rtol=2e-7 atol=2e-5
                    @test result.P_burnt_Pa == one_atm
                    # Cantera's default HP solve drifts by 8.5e-6 Pa here.
                    # Check the reference within its 1e-9 solver tolerance,
                    # while requiring exact prescribed pressure natively.
                    @test result.P_burnt_Pa ≈ Pref[i] rtol=1e-9
                    @test maximum(abs,Xnative-Xref[:,i]) < 1e-7
                    @test maximum(abs,result.Y_burnt-Yref[:,i]) < 1e-7
                else
                    @test all(isapprox.(Xnative,Xref[:,i];rtol=1e-12,atol=1e-14))
                    Ynative = Xnative .* gas.MW ./ dot(Xnative,gas.MW)
                    @test all(isapprox.(Ynative,Yref[:,i];rtol=1e-12,atol=1e-14))
                    Tnative = name in fresh ? result.T_fresh_K : result.T_post_burnt_K
                    @test Tnative ≈ Tref[i] rtol=2e-7 atol=2e-5
                    @test one_atm ≈ Pref[i] rtol=(name in fresh ? 1e-13 : 1e-9)
                end
                max_composition_error = max(max_composition_error,maximum(abs,Xnative-Xref[:,i]))
            end
        end
        for name in split(text("scalar_names_utf8"),'\n')
            @testset "$name" begin
                native = getproperty(result,Symbol(name))
                tolerance = name in ("phi_burnt","Z_burnt") ? 1e-8 : 1e-9
                @test native ≈ data["scalar_"*name] rtol=tolerance atol=1e-12
            end
        end
    end
    println("Maximum composition deviation: ",max_composition_error)
    return result
end

if abspath(PROGRAM_FILE) == @__FILE__
    length(ARGS) == 2 ||
        error("usage: julia validation/equivalence_ratio_case.jl <mechanisms_dir> <reference_dir>")
    validate_equivalence_ratio_case(ARGS[1],ARGS[2])
end

File: validation/equivalence_ratio_case.py
"""Exact pinned Cantera 4 thermo/equivalenceRatio.py reference generation.

https://github.com/Cantera/cantera/blob/726522be4e2a13454d8415b7ef799d621f665cf3/samples/python/thermo/equivalenceRatio.py
Reproduces every calculation in source order with the stateful gas object:
mixtures start at the default 300 K / 1 atm and the burnt temperature persists
for all mixtures created after equilibrate('HP') (called with source defaults).
Every recorded state (T, P, X, Y) and every printed scalar is stored in NPZ,
with metadata and mechanism hashes mirrored in JSON.

Run: python validation/equivalence_ratio_case.py mechanism output_directory
"""
from pathlib import Path
import sys
import hashlib
import json
import cantera as ct
import numpy as np

mechanisms, output = map(Path, sys.argv[1:3])
output.mkdir(parents=True, exist_ok=True)
if not ct.__version__.startswith("4."):
    raise RuntimeError("This reference requires Cantera 4")
mechanism = mechanisms / "gri30.yaml"
gas = ct.Solution(str(mechanism))
if gas.n_species != 53:
    raise RuntimeError("stock 53-species gri30.yaml required")

mixture_order, states, scalars = [], {}, {}


def record(name):
    mixture_order.append(name)
    states[name] = (gas.T, gas.P, gas.X.copy(), gas.Y.copy())


air = "O2:0.21,N2:0.79"

# Stoichiometric CH4/air mixture; keeps T = 300 K, P = 1 atm.
gas.set_equivalence_ratio(phi=1.0, fuel="CH4:1", oxidizer=air)
record("X_stoich_mole")

# Mass-basis fuel/oxidizer compositions.
gas.set_equivalence_ratio(1.0, fuel="CH4:1", oxidizer="O2:0.233,N2:0.767", basis="mass")
record("X_stoich_mass")

# Unnormalized amounts (233/767) normalize to the same oxidizer as above.
scalars["phi_mass_basis"] = gas.equivalence_ratio(fuel="CH4:1", oxidizer="O2:233,N2:767",
                                                  basis="mass")
scalars["phi_assumed_origin"] = gas.equivalence_ratio()
scalars["Z_default"] = gas.mixture_fraction(fuel="CH4:1", oxidizer=air)
scalars["Z_bilger"] = gas.mixture_fraction(fuel="CH4:1", oxidizer=air, element="Bilger")
scalars["Z_carbon"] = gas.mixture_fraction(fuel="CH4:1", oxidizer=air, element="C")
scalars["Y_CH4_stoich"] = gas["CH4"].Y[0]

gas.set_mixture_fraction(0.055, fuel="CH4:1", oxidizer=air)
record("X_Z055")
scalars["Y_CH4_Z055"] = gas["CH4"].Y[0]

fuel = {"CH4": 1}  # fuel composition as dictionary instead of string
gas.set_equivalence_ratio(1, fuel, air)
record("X_fresh_burnt_case")
gas.equilibrate("HP")
record("X_burnt")  # source T persists after this point
scalars["phi_burnt"] = gas.equivalence_ratio(fuel, air)
scalars["Z_burnt"] = gas.mixture_fraction(fuel, air)

fuel_arbitrary = "CH4:1,O2:0.01,CO:0.05,N2:0.1"
oxidizer_arbitrary = "O2:0.2,N2:0.8,CO2:0.05,CH4:0.01"
gas.set_equivalence_ratio(2.5, fuel=fuel_arbitrary, oxidizer=oxidizer_arbitrary)
record("X_arbitrary")
scalars["phi_arbitrary"] = gas.equivalence_ratio(fuel=fuel_arbitrary,
                                                 oxidizer=oxidizer_arbitrary)
scalars["phi_arbitrary_assumed"] = gas.equivalence_ratio()

gas.set_equivalence_ratio(2.0, "H2:1", "O2:1", diluent="H2O", fraction={"diluent": 0.3})
record("X_diluted_H2O")
scalars["X_H2O_diluted"] = gas["H2O"].X[0]
scalars["H2_O2_mole_ratio"] = gas["H2"].X[0] / gas["O2"].X[0]

gas.set_equivalence_ratio(2.0, "H2", "O2", diluent="CO2:0.5,H2O:0.5",
                          fraction={"fuel": 0.1}, basis="mass")
record("X_diluted_mass")
scalars["Y_H2_diluted"] = gas["H2"].Y[0]  # source prints this mass fraction
scalars["phi_include_species"] = gas.equivalence_ratio(fuel="H2", oxidizer="O2",
                                                       include_species=["H2", "O2"])

gas.set_equivalence_ratio(2.0, fuel="H2:0.5,H2O:0.5", oxidizer=air)
record("X_diluted_fuel")
scalars["phi_diluted_fuel"] = gas.equivalence_ratio(fuel="H2:0.5,H2O:0.5", oxidizer=air)

utf8 = lambda text: np.frombuffer(text.encode(), dtype=np.uint8)
mechanism_sha256 = hashlib.sha256(mechanism.read_bytes()).hexdigest()
arrays = {
    "X": np.stack([states[name][2] for name in mixture_order], axis=1),
    "Y": np.stack([states[name][3] for name in mixture_order], axis=1),
    "T_K": np.array([states[name][0] for name in mixture_order]),
    "P_Pa": np.array([states[name][1] for name in mixture_order]),
    "cantera_version_utf8": utf8(ct.__version__),
    "source_commit_utf8": utf8("726522be4e2a13454d8415b7ef799d621f665cf3"),
    "mechanism_sha256_utf8": utf8(mechanism_sha256),
    "mixture_names_utf8": utf8("\n".join(mixture_order)),
    "scalar_names_utf8": utf8("\n".join(scalars)),
    "species_names_utf8": utf8("\n".join(gas.species_names)),
}
arrays.update({f"scalar_{name}": value for name, value in scalars.items()})
np.savez(output / "equivalence_ratio-gri30.npz", **arrays)
manifest = dict(
    cantera=ct.__version__,
    source_commit="726522be4e2a13454d8415b7ef799d621f665cf3",
    source="samples/python/thermo/equivalenceRatio.py",
    mechanism=mechanism.name,
    mechanism_sha256=mechanism_sha256,
    species=gas.n_species,
    mixtures={name: {"T_K": states[name][0], "P_Pa": states[name][1]}
              for name in mixture_order},
    scalars={name: float(value) for name, value in scalars.items()},
)
(output / "equivalence_ratio-gri30.json").write_text(json.dumps(manifest, indent=2) + "\n")
print(json.dumps({"cantera": ct.__version__, "mixtures": mixture_order,
                  "scalars": manifest["scalars"]}, indent=2), flush=True)
